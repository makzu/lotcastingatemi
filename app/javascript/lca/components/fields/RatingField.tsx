import { ChangeEvent, ChangeEventHandler, useState, FocusEvent } from 'react'

import { TextField, TextFieldProps as MuiTextFieldProps } from '@mui/material'

import { useDebounce } from '@/hooks'
import { nuClamp as clamp } from '@/utils'

const selectOnFocus = (e: FocusEvent<HTMLInputElement>) => {
  e.target.select()
}

type RatingFieldProps = Omit<
  MuiTextFieldProps,
  'value' | 'children' | 'onChange'
> & {
  value: number
  /** @default 0 */
  min?: number
  /** @default Infinity */
  max?: number

  onChange: ChangeEventHandler<HTMLInputElement>
}

const RatingField = (props: RatingFieldProps) => {
  const { value, onChange, ...otherProps } = props
  const [localValue, setLocalValue] = useState<number | '-'>(value || 0)

  const min = props.min ?? 0
  const max = props.max ?? Infinity

  const debouncedOnChange = useDebounce(onChange, 500)

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if ((min < 0 && e.target.value === '-') || e.target.value === '0-') {
      setLocalValue('-')
      return
    }

    let value = Number(e.target.value)
    if (isNaN(value)) {
      value = 0
    }
    value = clamp(value, min, max)

    setLocalValue(value)
    debouncedOnChange(e)
  }
  return (
    <TextField
      {...otherProps}
      value={localValue}
      onChange={handleChange}
      inputMode="numeric"
      inputProps={{
        min,
        'aria-valuemin': min,
        max,
        'aria-valuemax': max,
        step: 1,
        pattern: '-?[0-9]*',
      }}
      onFocus={selectOnFocus}
      sx={{
        width: '4em',
        marginRight: 1,
        ...props.sx,
      }}
    />
  )
}

export default RatingField
