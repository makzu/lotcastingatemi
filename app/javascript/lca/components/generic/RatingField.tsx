import classnames from 'classnames'
import { Component, type SyntheticInputEvent } from 'react'

import { withStyles } from '@mui/styles'
import { clamp } from '@/utils/'

import { TextField } from '@mui/material'

const styles = (theme) => ({
  field: {
    marginRight: theme.spacing(),
    width: '4em',
  },
  narrow: {
    marginRight: theme.spacing(),
    width: '3em',
  },
})

interface Props {
  trait: string
  label: string
  value: number
  onChange: $TSFixMeFunction
  min?: number
  max?: number
  margin?: 'none' | 'dense' | 'normal'
  narrow?: boolean
  dontFocus?: boolean
  className?: string
  classes: Record<string, $TSFixMe>
}
interface State {
  value: number
  oldValue: number
} // TODO: Special fields for x/y resources like mote/willpower pools

class RatingField extends Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      value: this.props.value,
      oldValue: this.props.value,
    }
  }

  static defaultProps = {
    min: 0,
    max: Infinity,
  }
  static getDerivedStateFromProps = (props, state) => {
    if (state.oldValue === props.value) return null
    return {
      value: props.value,
      oldValue: props.value,
    }
  }

  handleChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { min, max } = this.props

    if (isNaN(parseInt(e.target.value))) {
      // $FlowThisIsOkayISwear
      this.setState({
        value: e.target.value,
      })
      return
    }

    // $FlowThisIsOkayISwear
    const value = clamp(parseInt(e.target.value), min, max)
    const fakeE = {
      target: {
        name: e.target.name,
        value: value,
      },
    }
    if (isNaN(parseInt(this.state.value)))
      this.setState({
        oldValue: this.state.value,
      })
    this.props.onChange(fakeE)
  }

  handleFocus = (e: SyntheticInputEvent<HTMLInputElement>) => {
    if (this.props.dontFocus) return
    e.target.select()
  }

  handleBlur = (e: SyntheticInputEvent<HTMLInputElement>) => {
    if (isNaN(parseInt(this.state.value))) {
      // $FlowThisIsOkayISwear
      this.setState({
        value: Math.max(0, this.props.min),
      })
      this.props.onChange({
        // $FlowThisIsOkayISwear
        target: {
          name: e.target.name,
          value: Math.max(0, this.props.min),
        },
      })
    }
  }

  render() {
    const {
      trait,
      label,
      classes,
      narrow,
      min,
      max,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      dontFocus,
      ...otherProps
    } = this.props
    const { handleChange, handleFocus, handleBlur } = this
    const { value } = this.state
    const className = classnames(
      narrow ? classes.narrow : classes.field,
      this.props.className,
    )
    return (
      <TextField
        {...otherProps}
        className={className}
        type="number"
        name={trait}
        label={label}
        value={value}
        inputProps={{
          min: min,
          max: max,
        }}
        onChange={handleChange}
        margin={this.props.margin || 'none'}
        onFocus={handleFocus}
        onBlur={handleBlur}
      />
    )
  }
}

export default withStyles(styles)(RatingField)
