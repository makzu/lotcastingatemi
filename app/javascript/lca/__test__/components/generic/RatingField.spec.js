// @flow
import renderer from 'react-test-renderer'

import RatingField from '../../../components/generic/RatingField.jsx'

test('RatingField renders at all', () => {
  const component = renderer.create(
    <RatingField trait="asdf" label="asdf" value={5} onChange={() => {}} />,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
