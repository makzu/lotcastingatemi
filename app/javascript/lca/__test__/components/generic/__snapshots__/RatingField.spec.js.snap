// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingField renders at all 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root RatingField-field-1"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
    data-shrink={true}
  >
    asdf
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiInput-input"
      disabled={false}
      max={Infinity}
      min={0}
      name="asdf"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="number"
      value={5}
    />
  </div>
</div>
`;
